zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+    /// The height increases as blocks are verified.
+    ///
+    /// If verification has not started yet, returns None.
+    /// If verification has finished, returns the maximum checkpoint height.
+    fn get_previous_checkpoint_height(&self) -> Option<BlockHeight> {
+        match self.current_checkpoint_range {
+            Some((Included(BlockHeight(0)), _)) => None,
+            Some((Excluded(height), _)) => Some(height),
+            None => self.get_max_checkpoint_height(),
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the next checkpoint height that we want to verify.
+    ///
+    /// The height increases as blocks are verified.
+    ///
+    /// If verification has not started yet, returns zero (the genesis block).
+    /// If verification has finished, returns None.
+    fn get_next_checkpoint_height(&self) -> Option<BlockHeight> {
+        match self.current_checkpoint_range {
+            Some((_, Included(height))) => Some(height),
+            None => None,
+            _ => unreachable!(),
+        }
+    }
+
+    /// Return the most recently verified checkpoint's hash.
+    ///
+    /// If verification has not started yet, returns None. To verify the genesis
+    /// block, check that the parent block hash is all zeroes. (Bitcoin "null".)
+    ///
+    /// If verification has finished, returns the maximum checkpoint's hash.
+    fn get_previous_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_previous_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
+            None => None,
+        }
+    }
+
+    /// Return the hash of the next checkpoint we want to verify.
+    ///
+    /// If verification has not started yet, returns the hash of block zero
+    /// (the genesis block). If verification has finished, returns None.
+    fn get_next_checkpoint_hash(&self) -> Option<BlockHeaderHash> {
+        match self.get_next_checkpoint_height() {
+            // Every checkpoint height must have a hash
+            Some(height) => Some(*self.checkpoint_list.get(&height).unwrap()),
+            None => None,
+        }
+    }
+
+    /// Return the height of the next checkpoint higher than `after_block_height`
+    /// in the checkpoint list. Ignores the current checkpoint range.
+    ///
+    /// If `after_block_height` is None, assume verification has not started yet,
+    /// and return zero (the genesis block).
+    ///
+    /// Returns None if there are no checkpoints. Also returns None if
+    /// `after_block_height` is greater than or equal to the maximum height.
+    fn find_descendant_checkpoint_height(
+        &self,
+        after_block_height: Option<BlockHeight>,
+    ) -> Option<BlockHeight> {
+        match after_block_height {
+            None => self.checkpoint_list.keys().cloned().next(),
+            Some(height) => self
+                .checkpoint_list
```

Comment:
TIOLI: you can call `.keys()` here and replace the map at the end with a `clone` I think.

Reply:
`BTreeMap::Range` doesn't implement `keys()`. And the `range()` is required to exclude completed checkpoints.